File: frontend/src/lib/api/attempts.ts
import { z } from "zod";
import { apiRequest } from "./client";

export type TimerMode = "COUNTDOWN" | "COUNT_UP";

export const attemptResponseSchema = z.object({
  attempt_id: z.string().uuid(),
  test_version_id: z.string().uuid(),
  test_session_id: z.string().uuid().nullable().optional(),
  attempt_context: z.enum(["STANDALONE", "FULL_MOCK"]).optional(),
  module: z.enum(["READING", "LISTENING", "WRITING"]),
  status: z.enum(["IN_PROGRESS", "PAUSED", "SUBMITTED", "AUTO_SUBMITTED", "INTERRUPTED", "ABANDONED"]),
  finished_reason: z.string().nullable(),
  timer_mode: z.enum(["COUNTDOWN", "COUNT_UP"]),
  timer_limit_seconds: z.number().nullable(),
  started_at: z.string(),
  paused_at: z.string().nullable(),
  total_paused_seconds: z.number().int().nonnegative(),
  deadline_at: z.string().nullable(),
  last_active_at: z.string(),
  finished_at: z.string().nullable(),
  elapsed_seconds: z.number(),
  remaining_seconds: z.number().nullable(),
  raw_score: z.number().nullable(),
  max_score: z.number().nullable(),
  band_score: z.number().nullable(),
  server_time: z.string(),
});

const writingResponseSchema = z.object({
  writing_task_id: z.string().uuid(),
  content: z.string(),
  word_count: z.number().int().nonnegative(),
  saved_at: z.string(),
});

export type AttemptResponse = z.infer<typeof attemptResponseSchema>;
export type WritingResponse = z.infer<typeof writingResponseSchema>;

export async function startAttempt(input: {
  test_version_id: string;
  module: "READING" | "LISTENING" | "WRITING";
  timer: { mode: TimerMode; duration_seconds?: number };
}) {
  return attemptResponseSchema.parse(await apiRequest<unknown>("/attempts", {
    method: "POST",
    body: JSON.stringify(input),
  }));
}

export function saveAnswer(attemptId: string, questionId: string, value: unknown) {
  return apiRequest(`/attempts/${attemptId}/answers/${questionId}`, {
    method: "PUT",
    body: JSON.stringify({ value }),
  });
}

export function recordActivity(attemptId: string) {
  return apiRequest(`/attempts/${attemptId}/activity`, {
    method: "POST",
    body: JSON.stringify({ client_observed_at: new Date().toISOString() }),
  });
}

export function recordNavigation(
  attemptId: string,
  kind: "PASSAGE" | "LISTENING_PART" | "WRITING_TASK" | "QUESTION",
  targetId: string,
) {
  return apiRequest(`/attempts/${attemptId}/navigation`, {
    method: "POST",
    body: JSON.stringify({ kind, target_id: targetId }),
  });
}

export async function saveWritingResponse(
  attemptId: string,
  writingTaskId: string,
  content: string,
): Promise<WritingResponse> {
  return writingResponseSchema.parse(
    await apiRequest<unknown>(`/attempts/${attemptId}/writing/${writingTaskId}`, {
      method: "PUT",
      body: JSON.stringify({ content }),
    }),
  );
}

export async function deleteAttempt(attemptId: string): Promise<void> {
  await apiRequest(`/attempts/${attemptId}`, { method: "DELETE" });
}

export async function pauseAttempt(attemptId: string): Promise<AttemptResponse> {
  return attemptResponseSchema.parse(
    await apiRequest<unknown>(`/attempts/${attemptId}/pause`, { method: "POST" }),
  );
}

export async function resumeAttempt(attemptId: string): Promise<AttemptResponse> {
  return attemptResponseSchema.parse(
    await apiRequest<unknown>(`/attempts/${attemptId}/resume`, { method: "POST" }),
  );
}

File: backend/app/schemas/attempts.py
from datetime import datetime
from decimal import Decimal
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, Field, field_validator, model_validator

from app.domains.scoring import validate_writing_criterion_score
from app.models.enums import (
    AttemptContext,
    AttemptStatus,
    FinishedReason,
    ModuleType,
    TestSessionStatus,
    TimerMode,
)
from app.schemas.assets import AssetResponse


class TimerRequest(BaseModel):
    mode: TimerMode
    duration_seconds: int | None = None

    @model_validator(mode="after")
    def validate_mode(self) -> "TimerRequest":
        presets = {2400, 3000, 3600, 4200}
        if self.mode == TimerMode.COUNTDOWN and self.duration_seconds not in presets:
            raise ValueError("Countdown duration must be 40, 50, 60, or 70 minutes")
        if self.mode == TimerMode.COUNT_UP and self.duration_seconds is not None:
            raise ValueError("Count-up mode does not accept a duration")
        return self


class AttemptCreate(BaseModel):
    test_version_id: UUID
    module: ModuleType
    timer: TimerRequest


class ActivityRequest(BaseModel):
    client_observed_at: datetime | None = None


class NavigationRequest(BaseModel):
    kind: Literal["PASSAGE", "LISTENING_PART", "WRITING_TASK", "QUESTION"]
    target_id: UUID


class AnswerUpdate(BaseModel):
    value: str | list[str] | dict[str, str]


class AnswerResponse(BaseModel):
    question_id: UUID
    value: Any
    is_correct: bool | None
    saved_at: datetime


class WritingResponseUpdate(BaseModel):
    content: str = Field(max_length=100_000)


class WritingResponse(BaseModel):
    writing_task_id: UUID
    content: str
    word_count: int = Field(ge=0)
    saved_at: datetime


class WritingTaskScoreUpdate(BaseModel):
    ta: Decimal
    cc: Decimal
    lr: Decimal
    gra: Decimal
    ta_feedback: str | None = Field(default=None, max_length=4000)
    cc_feedback: str | None = Field(default=None, max_length=4000)
    lr_feedback: str | None = Field(default=None, max_length=4000)
    gra_feedback: str | None = Field(default=None, max_length=4000)

    @field_validator("ta", "cc", "lr", "gra")
    @classmethod
    def validate_half_band(cls, value: Decimal) -> Decimal:
        return validate_writing_criterion_score(value)

    @field_validator("ta_feedback", "cc_feedback", "lr_feedback", "gra_feedback")
    @classmethod
    def normalize_feedback(cls, value: str | None) -> str | None:
        if value is None or not value.strip():
            return None
        return value.strip()


class WritingTaskScore(BaseModel):
    ta: float
    cc: float
    lr: float
    gra: float
    overall: float
    ta_feedback: str | None = None
    cc_feedback: str | None = None
    lr_feedback: str | None = None
    gra_feedback: str | None = None


class AttemptResponse(BaseModel):
    attempt_id: UUID
    test_version_id: UUID
    test_session_id: UUID | None = None
    attempt_context: AttemptContext = AttemptContext.STANDALONE
    module: ModuleType
    status: AttemptStatus
    finished_reason: FinishedReason | None
    timer_mode: TimerMode
    timer_limit_seconds: int | None
    started_at: datetime
    paused_at: datetime | None
    total_paused_seconds: int = Field(ge=0)
    deadline_at: datetime | None
    last_active_at: datetime
    finished_at: datetime | None
    elapsed_seconds: int
    remaining_seconds: int | None
    raw_score: int | None
    max_score: int | None
    band_score: float | None
    server_time: datetime


class ReviewAnswer(BaseModel):
    question_id: UUID
    question_number: int
    prompt: str
    value: Any
    answer_key: dict[str, Any]
    is_correct: bool | None
    explanation: str | None


class WritingReview(BaseModel):
    writing_task_id: UUID
    task_number: int
    prompt: str
    image_asset_id: UUID | None = None
    image_asset: AssetResponse | None = None
    minimum_recommended_words: int | None
    recommended_duration_seconds: int | None
    content: str
    word_count: int
    score: WritingTaskScore | None = None


class AttemptReview(BaseModel):
    attempt: AttemptResponse
    test_title: str
    answers: list[ReviewAnswer]
    writing_responses: list[WritingReview]
    highlights: list[dict[str, Any]]
    flags: list[dict[str, Any]]


class HistoryItem(BaseModel):
    attempt_id: UUID
    test_id: UUID
    test_version_id: UUID
    test_session_id: UUID | None = None
    test_title: str
    version_number: int
    module: ModuleType
    status: AttemptStatus
    started_at: datetime
    finished_at: datetime | None
    elapsed_seconds: int | None
    timer_mode: TimerMode
    timer_limit_seconds: int | None
    remaining_seconds: int | None
    raw_score: int | None
    max_score: int | None
    band_score: float | None
    review_available: bool = True


class HistoryGroup(BaseModel):
    test_id: UUID
    test_version_id: UUID
    test_title: str
    version_number: int
    reading: HistoryItem | None = None
    listening: HistoryItem | None = None
    writing: HistoryItem | None = None
    overall_band_score: float | None = None


class AttemptList(BaseModel):
    items: list[HistoryItem]
    groups: list[HistoryGroup]
    sessions: list["MockHistoryGroup"] = Field(default_factory=list)
    total: int = Field(ge=0)


class MockHistoryGroup(BaseModel):
    session_id: UUID
    test_version_id: UUID
    test_title: str
    version_number: int
    status: TestSessionStatus
    started_at: datetime
    finished_at: datetime | None
    reading: HistoryItem | None = None
    listening: HistoryItem | None = None
    writing: HistoryItem | None = None
    overall_band_score: float | None = None
